### PR TITLE
Implement pwm input for TIM2, TIM3, TIM4

### PR DESCRIPTION
**Pull request by gbip**
Hello !
I gave a try at implementing pwm input for the stm32f103, it was not tested (yet!) but the implementation is well documented.
Could you take a first look at it and tell me if you spot something wrong ?

Current features :
* Define which pins can be used for PwmInput
* Configure a timer in PwmInput mode
* Automate PRESC and ARR calculation
* Read a frequency or a duty cycle from a timer in PwmInput mode
* Implement PwmInput for TIM2, TIM3, TIM4

Thank you ! :smile:

### CODE REVIEW

**Comment by gbip** [comment]
cc @jujudu121212

**Comment by TheZoq2** [comment]
Thanks for the PR!

At a first glance, it looks very good. My only real complaint without looking into details about how the pwm capture works is that there are some lingering comments which would be nice if they were removed.

> it was not tested (yet!) but the implementation is well documented.

It would be nice if it was tested before we merge, are you able to do that?

**Comment by gbip** [comment]
I am looking forward testing it in the next few weeks.
I will also remove out of date comments !

**Comment by gbip** [comment]
So we spent a few hours testing this code with @jujudu121212 and there are a few problems :
* the struct `Clock` holds the frequency of the PLCK1/2 clocks, but ppre1 is used to multiply this frequency in some cases. However, the method to access `ppre1`/`ppre2` are privates.
* the code works for high frequencies (>600 kHz) but not for low frequencies... And we don't understand why... Do you have any idea ?

**Comment by eldruin** [comment]
Thanks for this @gbip!
From a quick glance, could you either remove commented-out code or include a comment about why they are commented out and what they do?

**Comment by gbip** [comment]
Hey !
This is the final version of our code, with @jujudu121212.
We tested with frequencies ranging from 1 Hz to 1 MHz an everything works fine.

Please tell me if there is any changes needed.

File: src/pwm_input.rs [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+            unsafe { mem::uninitialized() }
+         }
+
+      impl<PINS> PwmInput<$TIMX,PINS> where PINS : Pins<$TIMX> {
+         /// Return the frequency sampled by the timer
+         pub fn read_frequency(&self, mode : ReadMode, clocks : &Clocks) -> Result<Hertz,Error> {
+
+            match mode {
+               ReadMode::WaitForNextCapture => self.wait_for_capture(),
+                  _ => (),
```
Comment (TheZoq2) [inline src/pwm_input.rs]: Extra indentation

File: src/pwm_input.rs [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+            } else {
+               Ok((ccr2,ccr1))
+            }
+         }
+
+         /// Wait until the timer has captured a period 
+         fn wait_for_capture(&self) {
+            unsafe { (*$TIMX::ptr()).sr.write(|w| w.uif().clear_bit().cc1if().clear_bit().cc1of().clear_bit())};
+            while 
+            unsafe { (*$TIMX::ptr()).sr.read().cc1if().bit_is_clear()} {
```
Comment (TheZoq2) [inline src/pwm_input.rs]: Strange indentation and line break

File: src/rcc.rs [hunk 3]
```diff
@@ -325,13 +325,13 @@ impl Clocks {
         self.pclk2
     }
 
-    pub(crate) fn ppre1(&self) -> u8 {
+    pub fn ppre1(&self) -> u8 {
```
Comment (TheZoq2) [inline src/rcc.rs]: Why were these made public? As far as  I can tell, the are not used anywhere in this PR

File: src/gpio.rs [hunk 4]
```diff
@@ -311,19 +331,36 @@ macro_rules! gpio {
...
                                 w.bits((r.bits() & !(0b1111 << offset)) | (bits << offset))
                             });
 
-                        $PXi { _mode: PhantomData }
+                        res
                     }
-
-                    /// Configures the pin to operate as an push pull output pin
+                    /// Configures the pin to operate as an push-pull output pin.
+                    /// Initial state will be low.
```
Comment (TheZoq2) [inline src/gpio.rs]: Do we want to explicitly state the state of the pin in this function?

Also, it looks like the changes to this file aren't explicitly used in the PR, were they added by accident?

Comment (TheZoq2) [reply src/gpio.rs]: Well now i'm confused, these changes were part of #5 which is already merged, why are they showing up as diffs for this PR?

File: src/pwm_input.rs [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+        clocks: &Clocks,
+        mode: Configuration<T>,
+    ) -> PwmInput<Self, PINS>
+    where
+        PINS: Pins<Self>,
+        T: Into<Hertz>,
+    {
+        mapr.mapr()
+            .modify(|_, w| unsafe { w.tim3_remap().bits(PINS::REMAP) });
+        tim3(self, pins, clocks, apb, mode, dbg)
```
[the remark this answers is not in the document]
Comment (gbip) [reply src/pwm_input.rs]: `tim3(...)` is not a method, but an helper function (see [here](https://github.com/stm32-rs/stm32f1xx-hal/pull/10/files#diff-fbd3409c030ef50bd6dee46b7931eba6R212)).
The code is majorly inspired by the code in [`pwm.rs`](https://github.com/stm32-rs/stm32f1xx-hal/blob/master/src/pwm.rs#L143)

File: src/pwm_input.rs [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+/// <=> PRESC = PCLK / [ u16::MAX * F_timer ] -1
+fn compute_arr_presc(freq: u32, clock: u32) -> (u16, u16) {
+    let mut arr: u32 = (core::u16::MAX - 1) as u32;
+    let mut presc: u32;
+
+    let mut curr_best_presc : u32 = 0;
+    let mut curr_best_arr : u32 = 0;
+    let mut curr_best_distance : u32= core::u32::MAX;
+
+    for _ in 0..65530 {
```
[the remark this answers is not in the document]
Comment (gbip) [reply src/pwm_input.rs]: We are looking for a better algorithm, but this algorithm at least works.
Maybe we should mention in the documentation the fact that it is expensive to let the code find the best arr and presc value.

Before merging this PR we should also run a test were we try every combination possible.

File: src/pwm_input.rs [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+           if score < curr_best_distance {
+              curr_best_distance = score; 
+              curr_best_arr = arr;
+              curr_best_presc = presc;
+           }
+        }
+        arr -= 1;
+    }
+    assert!(curr_best_arr <= core::u16::MAX as u32);
+    assert!(curr_best_presc < core::u16::MAX as u32);
```
[the remark this answers is not in the document]
Comment (gbip) [reply src/pwm_input.rs]: I think that we will try to "prove" the correctness of this algorithm, so we could then remove those `assert!`.

File: src/pwm_input.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (jamwaffles) [reply src/pwm_input.rs]: Could you bisect (binary search) this instead? That'd only be 16 iterations at worst I think?

[the remark this answers is not in the document]
Comment (gbip) [reply src/pwm_input.rs]: Yes, this is what we plan to implement !

[the remark this answers is not in the document]
Comment (TheZoq2) [reply src/pwm_input.rs]: Makes sense, if it is linear, binary search does sound like a good idea, otherwise a slow algorithm with a comment mentioning its slowness sounds good enough for now imo

File: src/pwm_input.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (TheZoq2) [reply src/pwm_input.rs]: Ah right, I missed the declaration of that function

File: src/pwm_input.rs [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+            }
+
+            // Enable Capture on both channels
+            tim.ccer.modify(|_,w| w.cc1e().set_bit().cc2e().set_bit());
+
+
+            tim.cr1.modify(|_,w| w.cen().set_bit());
+
+
+            unsafe { mem::uninitialized() }
```
[the remark this answers is not in the document]
Comment (gbip) [reply src/pwm_input.rs]: This is the same behavior as for the [pwm](https://github.com/stm32-rs/stm32f1xx-hal/blob/master/src/pwm.rs) module.

As far as I understand, the idea is that the struct holds no data, and since everything is implemented through macros we can use `$TIMX` to have access to the pointer block.
I might be wrong because I am not an expert with `PhantomData` !

**Comment by gbip** [comment]
@TheZoq2 I have applied most of the changes you required.
I am sorry so many errors slept through my attention, I will be more careful next time.
For the automatic `presc` and `arr` calculation, maybe a comment saying that it is experimental and not yet validated is enough ?

I removed the asserts and used `Result` as a temporary measure, but it might break things once we are able to remove it.

I also plan to have a proofreading of all comments, since you may have noticed that English is not my first language.

File: src/pwm_input.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (TeXitoi) [reply src/pwm_input.rs]: What's the typical inputs for this function? I can search a better method.

File: src/pwm_input.rs [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+            let score = get_score(computed_freq, freq);
+            if score < curr_best_distance {
+                curr_best_distance = score;
+                curr_best_arr = arr;
+                curr_best_presc = presc;
+            }
+        }
+        arr -= 1;
+    }
+    if curr_best_arr <= core::u16::MAX as u32 || curr_best_presc < core::u16::MAX as u32 {
```
Comment (TeXitoi) [inline src/pwm_input.rs]: I think this condition is bad as you'll convert something over u16::MAX.

**Comment by TheZoq2** [comment]
> @TheZoq2 I have applied most of the changes you required.
I am sorry so many errors slept through my attention, I will be more careful next time.

Thanks, and no problem, I've submited much more unpolished PRs. This is what code review is for after all :)

> For the automatic `presc` and `arr` calculation, maybe a comment saying that it is experimental and not yet validated is enough?

I think that would be enough, but a result sounds like a better idea since that forces users to acknowlege that something can go wrong. That does mean that we'll have a breaking change when this is fixed which I don't consider a huge issue for such a small thing.

File: src/pwm_input.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (gbip) [reply src/pwm_input.rs]: Something between 1 and 12 000 000 for the frequency, and 8 000 000 up to 72 000 000 for the clock.

**Comment by gbip** [comment]
> > 
> 
> I think that would be enough, but a result sounds like a better idea since that forces users to acknowlege that something can go wrong. That does mean that we'll have a breaking change when this is fixed which I don't consider a huge issue for such a small thing.

Then if you don't mind grammar errors you can review this PR as I added the `Result` in my last commit !

File: src/pwm_input.rs (hunk 9 above)
Comment (gbip) [reply src/pwm_input.rs]: Thank you, this is now solved !

File: src/pwm_input.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (TeXitoi) [reply src/pwm_input.rs]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=515abfa4cd97d0274f66562bb0477851

[the remark this answers is not in the document]
Comment (TeXitoi) [reply src/pwm_input.rs]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8220a82864f38d668481f69ec49b3d1b

TLDR: can't fail:

```rust
fn compute_arr_presc(freq: u32, clock: u32) -> (u16, u16) {
    if freq == 0 {
        return (core::u16::MAX, core::u16::MAX);
    }
    let presc = clock / freq.saturating_mul(core::u16::MAX as u32 + 1);
    let mut arr = clock / freq.saturating_mul(presc + 1);
    if arr > 0 {
        arr -= 1;
    }
    (arr as u16, presc as u16)
}
```

[the remark this answers is not in the document]
Comment (TeXitoi) [reply src/pwm_input.rs]: If I understand correctly the RM 15.3.2 p.369

> n upcounting mode, the counter counts from 0 to the auto-reload value (content of the TIMx_ARR register), then restarts from 0 and generates a counter overflow event.

The formula is f = clock / (arr * (presc + 1)) (no `-1` to arr)

Thus the decrement in the code above is useless.

It's also said that arr = 0 will block the timer to not evoluting. That's what my function returns is freq > clock. Maybe returning `max(1, arr)` would be better.

So, this might be better:

```rust
fn compute_arr_presc(freq: u32, clock: u32) -> (u16, u16) {
    if freq == 0 {
        return (core::u16::MAX, core::u16::MAX);
    }
    let presc = clock / freq.saturating_mul(core::u16::MAX as u32 + 1);
    let arr = clock / freq.saturating_mul(presc + 1);
    (core::cmp::max(1, arr as u16), presc as u16)
}
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6cd00739ab7af5a763c598d52a3b3208

[the remark this answers is not in the document]
Comment (gbip) [reply src/pwm_input.rs]: Thank you @TeXitoi, I replaced our algorithm by yours !

File: src/pwm_input.rs [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+            let presc = unsafe { (*$TIMX::ptr()).psc.read().bits() as u16};
+            let ccr1 = unsafe { (*$TIMX::ptr()).ccr1.read().bits() as u16};
+
+            // Formulas :
+            //
+            // F_timer = F_pclk / (PSC+1)*(ARR+1)
+            // Frac_arr = (CCR1+1)/(ARR+1)
+            // F_signal = F_timer/Frac_arr
+            // <=> F_signal = [(F_plck)/((PSC+1)*(ARR+1))] * [(ARR+1)/(CCR1+1)]
+            // <=> F_signal = F_pclk / ((PSC+1)*(CCR1+1))
```
Comment (gbip) [inline src/pwm_input.rs]: Could someone check that our reasoning is correct ?

**Comment by gbip** [comment]
I think I addressed most of your requested changes, could you do a final pass on this PR ?

**Comment by nraynaud** [comment]
I think there is a oversight in the clock frequency computation.
The clock signal arriving to the timers might have already been pre-scaled:
clock tree

In [timer.rs](https://github.com/stm32-rs/stm32f1xx-hal/blob/8393fd0451d54f202f5295fa0409019c2bf4ad80/src/timer.rs#L125) they compute the frequency this way: 
```rust
                fn get_bus_frequency_multiplier(&self) -> u32 {
                    if TypeId::of::<$apbX>() == TypeId::of::<APB1>() {
                        if self.clocks.ppre1() == 1 {
                            1
                        } else {
                            2
                        }
                    } else if TypeId::of::<$apbX>() == TypeId::of::<APB2>() {
                         if self.clocks.ppre2() == 1 {
                            1
                         } else {
                            2
                         }
                    } else {
                        unreachable!()
                    }
                }
            }
```

in [pwm.rs](https://github.com/stm32-rs/stm32f1xx-hal/blob/8393fd0451d54f202f5295fa0409019c2bf4ad80/src/pwm.rs#L214) there is this bug (what if the timer is on APB2??):
```rust
let clk = clocks.pclk1().0 * if clocks.ppre1() == 1 { 1 } else { 2 };
```

so I suggest we just change [rcc.rs](https://github.com/nraynaud/stm32f103xx-hal/blob/nraynaud-changes/src/rcc.rs#L322) this way, and everybody uses this code:

```rust
impl Clocks {
....
    /// Returns the frequency of the APB1 Timers
    pub fn pclk1_tim(&self) -> Hertz {
        Hertz(self.pclk1.0 * if self.ppre1() == 1 { 1 } else { 2 })
    }

    /// Returns the frequency of the APB2 Timers
    pub fn pclk2_tim(&self) -> Hertz {
        Hertz(self.pclk2.0 * if self.ppre2() == 1 { 1 } else { 2 })
    }
....
}
```

**Comment by gbip** [comment]
Thank you !
We had identified this bug in our tests, this is why we made `ppr1e()` public and used it to compute clocks as a quick fix. This hacky fix got lost in my git problems.
Your solution is quite nicer, do you plan to make a PR ?

**Comment by nraynaud** [comment]
I think you are further along in your PR. I would try to sneak the RCC change along with your PR than go for another 15 days of code review

**Comment by gbip** [comment]
I preferred to make a PR this can wait a few more days.
Moreover since it is quite small #34 should get merged quickly !

**Comment by gbip** [comment]
Is there anything blocking this PR from being merged ? @TheZoq2

**Comment by TheZoq2** [comment]
Looks good to merge, thanks!
